paws_collector.js

```diff
@@ -263,6 +266,83 @@ class PawsCollector extends AlAwsCollector {
+        const collector = this;
+        const DDB = new AWS.DynamoDB();
+        const tableName = process.env.pawsDDBTableName;
+
+        const params = {
+            Key: {
+                "CollectorId": {S: collector._collectorId},
+                "MessageId": {S: stateSqsMsg.MD5OfBody}
+            },
+            TableName: tableName,
+            ConsistentRead: true
+        }
+
+        const getItemPromise = DDB.getItem(params).promise();
+
+        getItemPromise.then(data => {
+            // if the item is alread there, try and see if it is a duplicate
+            if (data.Item) {
+                // not sure if its right to call collector.done here
+                if(Item.Status.S === STATE_RECORD_INCOMPLETE && moment.unix() - Item.Updated.N < 900) {
+                    console.log(`Duplicate state: ${stateSqsMsg.MessageId}, already in progress. skipping`);
+                    return asyncCallback('State is currently being processed by another invocation');
+                } else if (Item.Status.S === STATE_RECORD_COMPLETE){
+                    console.log(`Duplicate state: ${stateSqsMsg.MessageId}, already processed. skipping`);
+                    return collector.done();
+                }
+                return collector.updateStateDBEntry(stateSqsMsg, STATE_RECORD_INCOMPLETE, asyncCallback);
+            // otherwise, put a new item in ddb.
+            } else {
+                const newRecord = {
+                    Item: {
+                        CollectorId: {S: collector._collectorId},
+                        MessageId: {S: stateSqsMsg.MessageId},
+                        Updated: {N: Date.now().toString()},
+                        Status: {S: STATE_RECORD_INCOMPLETE},
+                        // setting DDB time to life. This is the same as the sqs queue message retention
+                        ExpireDate: {N: moment().add(14, 'days').unix().toString()}
+                    },
+                    TableName: tableName
+                }
+                DDB.putItem(newRecord, (err) => {
+                    if(err){
+                        // TODO:figure out how to handle ddb write errors. Do they crash the collector?
+                    } else {
+                        return asyncCallback(null)
+                    }
+                });
+            }
+            
+        }).catch(asyncCallback)
+    }
+
+    updateStateDBEntry(stateSqsMsg, Status, asyncCallback) {
+        const collector = this;
+        const DDB = new AWS.DynamoDB();
+
+        const newRecord = {
+            Item: {
+                CollectorId: {S: collector._collectorId},
+                MessageId: {S: stateSqsMsg.MD5OfBody},
+                Updated: {N: Date.now().toString()},
+                Status: {S: Status},
+            },
+            TableName: tableName
+        }
+        DDB.putItem(newRecord, (err) => {
+            if(err){
+                // figure out how to handle ddb write errors
```

Comment:
we need to just fail the invocation so that SQS message is not consumed and retried.

Reply:
please return an error and `async.waterfall()` from `handleEvent` will handle it for you.